### PR TITLE
Make Job PodFailurePolicy e2e tests resilient to random failures to prepare them for conformance

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Follow up to the https://github.com/kubernetes/kubernetes/pull/125482#issuecomment-2233408132.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

This is what is currently implemented:
* drop the test "Job should allow to use the pod failure policy to not count the failure towards the backoffLimit" because it is redundant with "Job should allow to use a pod failure policy to ignore failure for an evicted pod; matching on the exit code"
* make the other tests resilient to the random failures, by using backoffLimit = completions, and triggering "completion" failures, which are all ignored. If they were not, the Job would fail.

Initially I tried this: use backoffLimit: 6 in the init phase (until all pods are running or succeeded), and reduce (update) backoffLimit to 0 (or the current number of Failures), then evict the pod with index 0, this way we get a single controlled failure. However. I realized, the Job may fail if the replacement pod for 0-indexed pod fails to schedule after reducing backoffLimit to 0.

Another idea is to watch that the UID of the evicted pod is not in 'uncountedTerminatedPods.Failed'. but this requires switching to non indexed jobs, and overall seems quite low level.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by mimowo** [comment]
/test pull-kubernetes-unit
Unrelated failure

**Comment by mimowo** [comment]
@atiratree @alculquicondor PTAL, I would like to move forward with this, so that we can promote the tests for Ignore action.

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go]: Thx, fixed

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go]: The reason was to make it more controlled and simpler to debug when I was developing the first version of the test.

Now, I think it is better to increase the parallelism to (1) decrease the test case runtime, (2) detect potential race conditions, so I pushed such a change now.

**Comment by mimowo** [comment]
FYI while doing the changes requested in some comments I also changed to things:
1. I noticed that I should use `backoffLimit = numPods - 1` to achieve the desired behavior, that the Job fails if the failures aren't ignored. Fixed. 
2. I use `numPods = 3 `rather than 2 to make the test more stable in case of random failures - the evicted Pods get ignored, so we can now tolerate up to 2 random failures

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Comment by alculquicondor** [comment]
/hold
in case @atiratree has anything to add

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go]: Thx, updated

File: test/e2e/apps/job.go
Comment (atiratree) [inline test/e2e/apps/job.go:169]: the comment bellow needs an update

File: test/e2e/apps/job.go
Comment (atiratree) [inline test/e2e/apps/job.go:186]:
```suggestion
		framework.ExpectNoError(err, "failed to await for all pods to be ready for job: %s/%s", job.Name, job.Namespace)
```

File: test/e2e/apps/job.go
Comment (atiratree) [inline test/e2e/apps/job.go:193]:
```suggestion
		ginkgo.By("Evict all the Pods")
```

;)

File: test/e2e/apps/job.go
Comment (atiratree) [reply test/e2e/apps/job.go:169]: also an observation: exit code 1 will not be observed. Is this intentional just to have the more values set? Might be usefulo to comment on that.

File: test/e2e/apps/job.go
Comment (atiratree) [inline test/e2e/apps/job.go:244]: same here; comments need an update

File: test/e2e/apps/job.go
Comment (atiratree) [inline test/e2e/apps/job.go:268]: same nits below in this test

File: test/e2e/apps/job.go
Comment (atiratree) [inline test/e2e/apps/job.go:214]: What is the reason of not checking the reason anymore?

**Review by atiratree — COMMENTED** [review]
just a bunch of comments need updating, the PR lgtm

> I use numPods = 3 rather than 2 to make the test more stable in case of random failures - the evicted Pods get ignored, so we can now tolerate up to 2 random failures. Looking at historical failures I've never seen more than 1 such a random failure in a single run (see https://github.com/kubernetes/kubernetes/pull/125482#issuecomment-2233408132), so we should be safe.

cool

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go:169]: I added exit code 1 as it was a potential scenario in the past: https://github.com/kubernetes/kubernetes/blob/20dd88872deb4516a6f8cbc32a59619951680c53/test/e2e/framework/job/fixtures.go#L157. However, we can simplify now. I will also delete `notTerminateOnce` as no longer used. It was more complex, so I'm happy we are able to get rid of it.

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go]: Thx

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go:244]: Done

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go:261]: done

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go]: done

**Comment by mimowo** [comment]
> /hold
> in case @atiratree has anything to add

/hold cancel
Since @atiratree provided review, and the new changes require lgtm anyway.

**Comment by mimowo** [comment]
/pull-kubernetes-unit  
Retest on unrelated failure

File: test/e2e/apps/job.go
Comment (atiratree) [reply test/e2e/apps/job.go:169]: okay, sounds good to me

File: test/e2e/apps/job.go
Comment (atiratree) [reply test/e2e/apps/job.go:214]: Hm, okay. But maybe we could add it to success policy graduation criteria, to start testing the reason again? WDYT? https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy#ga

**Comment by atiratree** [comment]
apart from a possible TODO in https://github.com/kubernetes/kubernetes/pull/126169#discussion_r1759182682
/lgtm

/hold

for @alculquicondor if he wants to double check the recent changes

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go:214]: Good idea, so that we don't forget about it, opened: https://github.com/kubernetes/enhancements/pull/4851

**Comment by mimowo** [comment]
/assign @pohly 
For `test/e2e/framework/OWNERS`

**Review by pohly — APPROVED** [review]
/approve

I trust the reviewers of test/e2e/framework/job/fixtures.go - it's not a file that I have personal experience with.

**Comment by mimowo** [comment]
> /hold
> for @alculquicondor if he wants to double check the recent changes

PTAL @alculquicondor

**Review by soltysh — APPROVED** [review]
Minor nit
/approve
/label tide/merge-method-squash

File: test/e2e/apps/job.go
Comment (mimowo) [reply test/e2e/apps/job.go]: Done, thx

**Comment by mimowo** [comment]
/test pull-kubernetes-unit
Unrelated flake as this PR only updates e2e tests

File: test/e2e/apps/job.go
Comment (pohly) [inline test/e2e/apps/job.go:205]: All goroutines use the same `err` variable -> data race.

```
==================
  WARNING: DATA RACE
  Write at 0x00c000cfe250 by goroutine 21297:
    k8s.io/kubernetes/test/e2e/apps.init.func4.4.1()
        k8s.io/kubernetes/test/e2e/apps/job.go:259 +0x718
    k8s.io/kubernetes/test/e2e/apps.init.func4.4.1()
        k8s.io/kubernetes/test/e2e/apps/job.go:259 +0x6fe
    k8s.io/client-go/util/workqueue.ParallelizeUntil.func1()
        k8s.io/client-go/util/workqueue/parallelizer.go:90 +0x218

  Previous read at 0x00c000cfe250 by goroutine 21298:
    k8s.io/kubernetes/test/e2e/apps.init.func4.4.1()
        k8s.io/kubernetes/test/e2e/apps/job.go:260 +0x75c
    k8s.io/kubernetes/test/e2e/apps.init.func4.4.1()
        k8s.io/kubernetes/test/e2e/apps/job.go:259 +0x6fe
    k8s.io/client-go/util/workqueue.ParallelizeUntil.func1()
        k8s.io/client-go/util/workqueue/parallelizer.go:90 +0x218

  Goroutine 21297 (running) created at:
    k8s.io/client-go/util/workqueue.ParallelizeUntil()
        k8s.io/client-go/util/workqueue/parallelizer.go:76 +0x2d1
    k8s.io/kubernetes/test/e2e/apps.init.func4.4()
        k8s.io/kubernetes/test/e2e/apps/job.go:244 +0xf47
    k8s.io/kubernetes/test/e2e/apps.init.func4.4()
        k8s.io/kubernetes/test/e2e/apps/job.go:235 +0x7b8
    github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func2()
        github.com/onsi/ginkgo/v2@v2.27.2/internal/node.go:517 +0x5d
    github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
        github.com/onsi/ginkgo/v2@v2.27.2/internal/suite.go:942 +0x6ed

  Goroutine 21298 (finished) created at:
    k8s.io/client-go/util/workqueue.ParallelizeUntil()
        k8s.io/client-go/util/workqueue/parallelizer.go:76 +0x2d1
    k8s.io/kubernetes/test/e2e/apps.init.func4.4()
        k8s.io/kubernetes/test/e2e/apps/job.go:244 +0xf47
    k8s.io/kubernetes/test/e2e/apps.init.func4.4()
        k8s.io/kubernetes/test/e2e/apps/job.go:235 +0x7b8
    github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func2()
        github.com/onsi/ginkgo/v2@v2.27.2/internal/node.go:517 +0x5d
    github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
        github.com/onsi/ginkgo/v2@v2.27.2/internal/suite.go:942 +0x6ed
  ==================
```

Comment (soltysh) [reply test/e2e/apps/job.go:205]: Addressed in https://github.com/kubernetes/kubernetes/pull/135610

**Pull request merged**
